File: bin/sp-4.4.1/sicstus-4.4.1/se/sics/prologbeans/PBCompound.java
/**
 * Copyright (c) 2003 SICS AB. All rights reserved.
 */
package se.sics.prologbeans;

import java.io.IOException;

/**
 * <code>PBCompound</code> is the
 * representation of Prolog compound terms.
 */
class PBCompound extends PBTerm {
	protected final PBTerm[] arguments;

	PBCompound(String name, PBTerm[] args) {
		super(name);
		this.arguments = args;
	}

	public boolean isAtom() {
		return getArity() == 0;
	}

	public boolean isAtomic() {
		return getArity() == 0;
	}

	public boolean isListCell() {
		return getArity() == 2 && ".".equals(getName());
	}

	public boolean isProperList() {
		// return (isListCell() && getArgument(2).isProperList())
		// || isEmptyList();
		return (isListCell() && getArgument(2).isProperList());
	}

	public PBTerm getArgument(int index) {
		if (arguments == null) {
			throw new IllegalStateException("not a compound term: "
					+ toString());
		}
		if (index < 1 || index > arguments.length) {
			throw new IndexOutOfBoundsException("Index: " + index
					+ ", needs to be between 1 and " + getArity());
		}
		return arguments[index - 1];
	}

	public int getArity() {
		return arguments == null ? 0 : arguments.length;
	}

	String toPrologString() {
		StringBuffer sb = new StringBuffer().append(stuffAtom(name));
		if (arguments != null) {
			sb.append('(');
			for (int i = 0, n = arguments.length; i < n; i++) {
				if (i > 0) {
					sb.append(',');
				}
				sb.append(arguments[i].toPrologString());
			}
			sb.append(')');
		}
		return sb.toString();
	}

	public String toString() {
		StringBuffer sb = new StringBuffer().append(name);
		if (arguments != null) {
			sb.append('(');
			for (int i = 0, n = arguments.length; i < n; i++) {
				if (i > 0) {
					sb.append(',');
				}
				sb.append((arguments[i] != null ? arguments[i].toString()
						: "<<NULL>>"));
			}
			sb.append(')');
		}
		return sb.toString();
	}

	void fastWrite(FastWriter writer) throws IOException {
		if (arguments != null) {
			writer.writeCompound(name, arguments.length);
			for (int i = 0, n = arguments.length; i < n; i++) {
				arguments[i].fastWrite(writer);
			}
		} else {
			throw new IllegalStateException("not a compound term: "
					+ toString());
		}
	}

	// @Override
	PBTerm[] fastWritePrefix(FastWriter writer) throws IOException {
		if (arguments != null) {
			writer.writeCompound(name, arguments.length);
			return arguments;
		} else {
			throw new IllegalStateException("not a compound term: "
					+ toString());
		}
	}

}

File: bin/sp-4.4.1/sicstus-4.4.1/library/prologbeans.NET/PBCompound.cs
// Copyright (c) 2004 SICS AB. All rights reserved.
// 
namespace se.sics.prologbeans
{
	using System;

	/// <summary> <code>PBCompound</code> is the
	/// representation of Prolog compound terms.
	/// </summary>
	class PBCompound:PBTerm
	{
		override public sealed bool Compound {
			get {
				return true;
			}

		}

		override public int Arity {
			get {
				return arguments.Length;
			}
		}

		protected internal readonly PBTerm[] arguments;

		// Must never be used to create a (.)/2 term (i.e. a list cell).
		internal PBCompound (string name, PBTerm[] args) : base (name)
		{
			// assert (args != null && args.Length > 0 && !(".".equals(name) && args.Length == 2))
			arguments = args;
		}

		public override PBTerm getArgument (int index)
		{
			if (index < 1 || index > arguments.Length) {
				throw new ArgumentOutOfRangeException ("index");
			}
			return arguments [index - 1];
		}

		internal override string toPrologString ()
		{
			// [PM] 4.3.2 FIXME: stack overflow issue.
			System.Text.StringBuilder sb = new System.Text.StringBuilder ().Append (stuffAtom (Name));

			sb.Append ('(');
			for (int i = 0, n = arguments.Length; i < n; i++) {
				if (i > 0) {
					sb.Append (',');
				}
				sb.Append (arguments [i].toPrologString ());
			}
			sb.Append (')');

			return sb.ToString ();
		}

		public override string ToString ()
		{
			// [PM] 4.3.2 FIXME: stack overflow issue.
			System.Text.StringBuilder sb = new System.Text.StringBuilder ().Append (Name);
		
			sb.Append ('(');
			for (int i = 0, n = arguments.Length; i < n; i++) {
				if (i > 0) {
					sb.Append (',');
				}
				sb.Append ((arguments [i] != null ? arguments [i].ToString () : "<<NULL>>"));
			}
			sb.Append (')');

			return sb.ToString ();
		}

		internal override PBTerm[] fastWritePrefix (FastWriter writer)
		{
			writer.writeCompound (Name, arguments.Length);
			return arguments;
		}
	}
}
